lib/internal/encoding.js

```diff
@@ -0,0 +1,359 @@
+function hasTextDecoder(encoding = 'utf-8') {
+  if (typeof encoding !== 'string')
+    throw new errors.Error('ERR_INVALID_ARG_TYPE', 'encoding', 'string');
+  return hasConverter(normalizeEncoding(encoding));
+}
+
+class TextDecoder {
+  constructor(encoding = 'utf-8', options = {}) {
+    if (typeof encoding !== 'string')
+      throw new errors.Error('ERR_INVALID_ARG_TYPE', 'encoding', 'string');
+    if (options !== undefined && typeof options !== 'object')
+      throw new errors.Error('ERR_INVALID_ARG_TYPE', 'options', 'object');
+
+    const fatal = options !== null &&
+                  options.fatal === true;
+
+    const ignoreBOM = options !== null &&
+                      options.ignoreBOM === true;
+
+    const enc = normalizeEncoding(encoding);
+    if (enc === undefined || enc === 'replacement')
+      throw new errors.RangeError('ERR_ENCODING_NOT_SUPPORTED', encoding);
+
+    var flags = 0;
+    flags |= fatal ? CONVERTER_FLAGS_FATAL : 0;
+    flags |= ignoreBOM ? CONVERTER_FLAGS_IGNORE_BOM : 0;
+
+    const handle = getConverter(enc, flags);
+    if (handle === undefined)
+      throw new errors.Error('ERR_ENCODING_NOT_SUPPORTED', encoding);
+
+    Object.defineProperties(this, {
+      [kHandle]: {
+        configurable: false,
+        enumerable: false,
+        value: handle
+      },
+      encoding: {
+        configurable: false,
+        enumerable: true,
+        value: enc
+      },
+      fatal: {
+        configurable: false,
+        enumerable: true,
+        value: fatal
+      },
+      ignoreBOM: {
+        configurable: false,
+        enumerable: true,
+        value: ignoreBOM
+      }
+    });
+  }
+
+  decode(input, options = {}) {
+    if (input instanceof ArrayBuffer) {
+      input = Buffer.from(input);
+    } else if (!ArrayBuffer.isView(input)) {
+      throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'input',
+                                 ['ArrayBuffer', 'ArrayBufferView']);
+    }
+    if (options === null ||
+        typeof options !== 'object') {
+      throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'options', 'object');
+    }
+
+    var flags = 0;
+    flags |= (options.stream === true) ? 0 : CONVERTER_FLAGS_FLUSH;
+
+    const ret = _decode(this[kHandle], input, flags);
+    if (typeof ret === 'number') {
+      const err = new errors.TypeError('ERR_ENCODING_INVALID_ENCODED_DATA',
+                                       this.encoding);
+      err.errno = ret;
+      throw err;
+    }
+    return ret.toString('ucs2');
+  }
+}
+
+var Buffer;
+function lazyBuffer() {
+  if (Buffer === undefined)
+    Buffer = require('buffer').Buffer;
+  return Buffer;
+}
+
+class TextEncoder {
+  constructor() {
+    Object.defineProperties(this, {
+      encoding: {
+        configurable: false,
+        enumerable: true,
+        value: 'utf-8'
+      }
+    });
+  }
+
+  encode(input) {
+    const buf = lazyBuffer().from(String(input));
+    return new Uint8Array(buf.buffer, buf.offset, buf.length);
```

Comment:
Per [spec](https://encoding.spec.whatwg.org/#dom-textencoder-encode), the returned Uint8Array must be unpooled and have an entire block of ArrayBuffer dedicated to it.